### PR TITLE
ENT-3088 | Added new view for requesting the DSC for learners for spe…

### PR DESCRIPTION
**Pull request by hasnain-naveed**
…cific course.

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [Travis](https://travis-ci.org/github/edx/edx-enterprise), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

**Comment by hasnain-naveed** [comment]
You have to test the following use cases:

**Manage Learner View:**

1. Link user.
2. Link User by adding course, reason and SF id.
3. Link bulk user by adding CSV file

**Transmit Meta data**
Add a worker user and see if command got the hit

**Manage Learners DSC**
1) Add a valid user (linked to ENT) and valid course (Exist in ENT catalog) then its DSC record should be removed and it's cached should also be destroyed. 

Let me know if you have any more question.

File: enterprise/admin/views.py [hunk 1]
```diff
@@ -133,43 +134,161 @@ def _build_admin_context(request, customer):
...
+        """
+        Check whether user is linked to the enterprise customer or not.
+        """
+        user = User.objects.get(email=email)
+        return get_enterprise_customer_user(user.id, customer_uuid)
+
+    @staticmethod
+    def is_course_in_catalog(course_id, customer_uuid):
+        """
+        Check whether user is linked to the enterprise customer or not.
```
[the remark this answers is not in the document]
Comment (sameenfatima78) [reply enterprise/admin/views.py:189]: Nit: PR description has the wrong ticket number.

**Comment by sameenfatima78** [comment]
When submitting the `Transmit Courses Metadata` form with enterprise worker username: `enterprise_worker` it redirects to a blank page. The command does get hit in the post endpoint but it is returning a blank page. Is this the required behaviour or am I missing something?

File: enterprise/admin/views.py [hunk 2]
```diff
@@ -133,43 +134,161 @@ def _build_admin_context(request, customer):
...
-        return render(request, self.template, context)
+    def post(self, request, customer_uuid):
+        """
+        Handle POST request - handle form submissions.
+
+        Arguments:
+            request (django.http.request.HttpRequest): Request instance
+            customer_uuid (str): Enterprise Customer UUID
+        """
+        manage_learners_dsc = ManageLearnersDataSharingConsentForm(request.POST)
```
Comment (saleem-latif) [inline enterprise/admin/views.py:221]: nit: rename it to `manage_learners_dsc_form`

File: enterprise/admin/views.py [hunk 3]
```diff
@@ -133,43 +134,161 @@ def _build_admin_context(request, customer):
...
+
+        # check that form data is well-formed
+        if manage_learners_dsc.is_valid():
+            user_email = manage_learners_dsc.cleaned_data[ManageLearnersDataSharingConsentForm.Fields.EMAIL_OR_USERNAME]
+            course_id = manage_learners_dsc.cleaned_data[ManageLearnersDataSharingConsentForm.Fields.COURSE]
+
+            if not self.is_user_linked(user_email, customer_uuid):
```
Comment (saleem-latif) [inline enterprise/admin/views.py:228]: I think this logic should be moved to the form's clean method.

File: enterprise/admin/views.py [hunk 4]
```diff
@@ -133,43 +134,161 @@ def _build_admin_context(request, customer):
...
+                manage_learners_dsc.add_error(
+                    ManageLearnersDataSharingConsentForm.Fields.EMAIL_OR_USERNAME, ValidationMessages.USER_NOT_LINKED
+                )
+            elif not self.is_course_in_catalog(course_id, customer_uuid):
```
Comment (saleem-latif) [inline enterprise/admin/views.py:232]: Same as above, I think this logic should be moved to the form's clean method.

File: enterprise/admin/views.py [hunk 5]
```diff
@@ -212,26 +335,8 @@ class ContextParameters:
...
-            "has_change_permission": has_change_permission,
-            "opts": opts
-        }
-
-    def _build_context(self, request, customer_uuid):
-        """
-        Build common context parts used by different handlers in this view.
-        """
-        # TODO: pylint acts stupid - find a way around it without suppressing
+    def _built_view_context(self, request, customer_uuid):
```
Comment (saleem-latif) [inline enterprise/admin/views.py:338]: nit: `_get_view_context` seems more appropriate

File: enterprise/admin/views.py [hunk 6]
```diff
@@ -767,14 +870,16 @@ def get(self, request, customer_uuid):
         Returns:
             django.http.response.HttpResponse: HttpResponse
         """
-        context = self._build_context(request, customer_uuid)
-        manage_learners_form = ManageLearnersForm(
-            user=request.user,
-            enterprise_customer=context[self.ContextParameters.ENTERPRISE_CUSTOMER]
-        )
-        context.update({self.ContextParameters.MANAGE_LEARNERS_FORM: manage_learners_form})
+        enterprise_customer = EnterpriseCustomer.objects.get(uuid=customer_uuid)  # pylint: disable=no-member
```
Comment (saleem-latif) [inline enterprise/admin/views.py:873]: are we sure this will never raise `EnterpriseCustomer.DoesNotExist` error?

Comment (hasnain-naveed) [reply enterprise/admin/views.py:873]: Yes.

File: enterprise/admin/views.py [hunk 7]
```diff
@@ -133,43 +134,161 @@ def _build_admin_context(request, customer):
...
         """
-        Build common context parts used by different handlers in this view.
+        Return the default context parameters
         """
-        enterprise_customer = EnterpriseCustomer.objects.get(uuid=enterprise_customer_uuid)  # pylint: disable=no-member
-
-        context = {
-            self.ContextParameters.ENTERPRISE_CUSTOMER: enterprise_customer,
+        return {
+            'enterprise_customer': EnterpriseCustomer.objects.get(uuid=customer_uuid)  # pylint: disable=no-member
```
[the remark this answers is not in the document]
Comment (hasnain-naveed) [reply enterprise/admin/views.py:142]: yes

File: enterprise/admin/forms.py [hunk 8]
```diff
@@ -199,11 +198,12 @@ def clean_course(self):
         course_id = self.cleaned_data[self.Fields.COURSE].strip()
         if not course_id:
             return None
-        try:
-            client = EnrollmentApiClient()
-            return client.get_course_details(course_id)
-        except (HttpClientError, HttpServerError):
+
+        client = EnrollmentApiClient()
+        course_details = client.get_course_details(course_id)
```
[the remark this answers is not in the document]
Comment (hasnain-naveed) [reply enterprise/admin/forms.py:203]: Because this is being handled in `get_course_details` function.

**Comment by hasnain-naveed** [comment]
@sameenfatima78 yes that is the acceptable behavior. If you follow the code you will see that it is redirecting to the empty page.

File: enterprise/admin/forms.py [hunk 9]
```diff
@@ -280,6 +280,47 @@ def _validate_reason(self):
...
+            raise ValidationError(ValidationMessages.INVALID_COURSE_ID.format(course_id=course_id))
+        return course_id
+
+    def clean_email_or_username(self):
+        """
+        Verify email_or_username has associated user in our database.
+        """
+        email_or_username = self.cleaned_data[self.Fields.EMAIL_OR_USERNAME].strip()
+        email = email_or_username__to__email(email_or_username)
+        if not User.objects.filter(email=email).exists():
```
Comment (zamanafzal) [inline enterprise/admin/forms.py:319]: We are only filtering the user here on the basis of email. Then why we are taking the username?

Comment (hasnain-naveed) [reply enterprise/admin/forms.py:319]: `email_or_username__to__email ` function will convert the username to email.

**Pull request merged**